huaweicloud/resource_huaweicloud_cbr_vault.go

```diff
@@ -0,0 +1,431 @@
+		return fmt.Errorf("Error deleting CBR v3 vault: %s", err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourceIncludeVolume(volumes []interface{}) []vaults.ResourceExtraInfoIncludeVolumes {
+	includeVolumes := make([]vaults.ResourceExtraInfoIncludeVolumes, len(volumes))
+	if len(volumes) == 0 {
+		return includeVolumes
+	}
+	for i, v := range volumes {
+		includeVolumes[i] = vaults.ResourceExtraInfoIncludeVolumes{
+			ID: v.(string),
+		}
+	}
+	return includeVolumes
+}
+
+func resourceExcludeVolume(volumes []interface{}) []string {
+	includeVolumes := make([]string, len(volumes))
+	if len(volumes) == 0 {
+		return includeVolumes
+	}
+	for i, v := range volumes {
+		includeVolumes[i] = v.(string)
+	}
+	return includeVolumes
+}
+
+func buildCBRVaultResources(d *schema.ResourceData) []vaults.ResourceCreate {
+	res := make([]vaults.ResourceCreate, 0)
+	resType := d.Get("type").(string)
+	for _, v := range d.Get("resources").(*schema.Set).List() {
+		resourceID := v.(map[string]interface{})["id"].(string)
+		resourceType := resourceType[resType]
+		includes := resourceIncludeVolume(v.(map[string]interface{})["include_volumes"].([]interface{}))
+		excludes := resourceExcludeVolume(v.(map[string]interface{})["exclude_volumes"].([]interface{}))
+		extra := vaults.ResourceExtraInfo{
+			IncludeVolumes: includes,
+			ExcludeVolumes: excludes,
+		}
+		res = append(res, vaults.ResourceCreate{
+			ID:        resourceID,
+			Type:      resourceType,
+			ExtraInfo: &extra,
+		})
+	}
+	return res
+}
+
+func buildCBRVaultBilling(d *schema.ResourceData) *vaults.BillingCreate {
+	billing := &vaults.BillingCreate{
+		ObjectType:      d.Get("type").(string),
+		ConsistentLevel: d.Get("consistent_level").(string),
+		ProtectType:     d.Get("protection_type").(string),
+		Size:            d.Get("size").(int),
+	}
+
+	return billing
+}
+
+//Take out and save all the IDs nested in the list of resources map as a new string array.
+func changeSliceElementType(slice []interface{}) []string {
+	result := make([]string, len(slice))
+	for i, v := range slice {
+		resMap := v.(map[string]interface{})
+		result[i] = resMap["id"].(string)
+	}
+	return result
+}
+
+//Take out and save all the IDs and resource type nested in the list of resources map as a new ResourceCreate array.
+func buildResourcesSlice(slice []interface{}, resType string) []vaults.ResourceCreate {
+	result := make([]vaults.ResourceCreate, len(slice))
+	for i, v := range slice {
+		resMap := v.(map[string]interface{})
+		result[i] = vaults.ResourceCreate{
+			ID:   resMap["id"].(string),
+			Type: resType,
+		}
+	}
+	return result
+}
+
+func updateResources(d *schema.ResourceData, client *golangsdk.ServiceClient) error {
+	oldRaws, newRaws := d.GetChange("resources")
+	resType := d.Get("type").(string)
+
+	//Calculate the intersection of oldRaws and newRaws, and use the intersection to calculate the subtraction.
+	intersectionRaws := oldRaws.(*schema.Set).Intersection(newRaws.(*schema.Set))
```

Comment:
maybe the line is unnecessary

Reply:
These codes have been updated according to other resource codes.